packages/wallet/src/services/ChangeAddress/DynamicChangeAddressResolver.ts

```diff
@@ -0,0 +1,369 @@
+ * rigorous approach such as integer programming, simulated annealing, or a complete search, which could examine all possible
+ * assignments of change outputs to buckets and select the one that most closely matches the desired proportions. However,
+ * these methods can be computationally intensive and may not be practical, and probably overkill for our use case (To be determined).
+ *
+ * @param changeOutputs The list of change outputs to be distributed.
+ * @param prefilledBuckets The list of buckets, we should have one bucket per entry in the delegation distribution portfolio.
+ * @returns A list of buckets with the given change outputs distributed in a way that closely matches the expected distribution (best effort).
+ */
+const distributeChange = (changeOutputs: Array<Cardano.TxOut>, prefilledBuckets: Array<Bucket>): Array<Bucket> => {
+  const buckets = [...prefilledBuckets];
+  const sortedOutputs = changeOutputs.sort((a, b) => {
+    if (a.value.coins > b.value.coins) {
+      return 1;
+    } else if (a.value.coins < b.value.coins) {
+      return -1;
+    }
+    return 0;
+  });
+
+  while (sortedOutputs.length > 0) {
+    const bucket = pickBucket(buckets);
+    const selected = sortedOutputs.splice(0, 1)[0];
+    bucket.change.push(selected);
+    bucket.filledAmount += selected.value.coins;
+  }
+
+  for (const bucket of buckets) {
+    bucket.change = bucket.change.map((txOut) => {
+      txOut.address = bucket.address;
+      return txOut;
+    });
+  }
+
+  return buckets;
+};
+
+/**
+ * Gets whether the portfolio pools matches the current stake distribution pools.
+ *
+ * @param portfolio The staking portfolio.
+ * @param distribution the distribution.
+ * @returns true if both matches, otherwise; false.
+ */
+const delegationMatchesPortfolio = (
+  portfolio: Cardano.Cip17DelegationPortfolio,
+  distribution: DelegatedStake[]
+): boolean => {
+  const portfolioPools = uniq(portfolio.pools.map((cip17Pool) => cip17Pool.id)).sort();
+  const delegationPools = uniq(distribution.map((delegatedStake) => delegatedStake.pool.hexId)).sort();
+
+  return isEqual(portfolioPools, delegationPools);
+};
+
+/**
+ * Gets the current delegation portfolio.
+ */
+export type GetDelegationPortfolio = () => Promise<Cardano.Cip17DelegationPortfolio | null>;
+
+/**
+ * Resolves the address to be used for change.
+ */
+export class DynamicChangeAddressResolver implements ChangeAddressResolver {
+  readonly #getDelegationPortfolio: GetDelegationPortfolio;
+  readonly #delegationTracker: DelegationTracker;
+  readonly #addresses$: Observable<GroupedAddress[]>;
+
+  /**
+   * Initializes a new instance of the StakeDistributionChangeAddressResolver.
+   *
+   * @param addresses$ The wallet known addresses.
+   * @param delegationTracker The wallet delegation tracker.
+   * @param getDelegationPortfolio The current delegation portfolio.
+   */
+  constructor(
+    addresses$: Observable<GroupedAddress[]>,
+    delegationTracker: DelegationTracker,
```

Comment:
Consider narrowing this down
```suggestion
    distribution$: DelegationTracker['distribution$'],
```